File: app/src/main/java/com/soutvoid/gamesproject/ui/screen/main/widgets/exploreset/widget/ExploreSetView.java
package com.soutvoid.gamesproject.ui.screen.main.widgets.exploreset.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.soutvoid.gamesproject.domain.game.Game;
import com.soutvoid.gamesproject.interactor.util.ImageUrlBuilder;
import com.soutvoid.gamesproject.ui.screen.main.widgets.exploreset.list.ExploreSetListAdapter;
import com.soutvoid.gamesproject.ui.util.OnGameClickListener;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import lombok.Setter;
import soutvoid.com.gamesproject.R;

public class ExploreSetView extends FrameLayout {

    private Context context;

    @BindView(R.id.main_explore_set_container)
    ViewGroup container;
    @BindView(R.id.main_explore_set_background)
    ImageView background;
    @BindView(R.id.main_explore_set_background_tint)
    View backgroundTint;
    @BindView(R.id.main_explore_set_header)
    TextView header;
    @BindView(R.id.main_explore_set_list)
    RecyclerView list;

    private ExploreSetListAdapter exploreSetListAdapter;

    @Setter
    private OnGameClickListener onGameClickListener;

    public ExploreSetView(Context context) {
        super(context);
        init(context, null);
    }

    public ExploreSetView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    private void init(Context context, @Nullable AttributeSet attrs) {
        this.context = context;
        inflate(context, R.layout.explore_set_view, this);
        ButterKnife.bind(this);
        fillFromAttrs(context, attrs);
    }

    private void fillFromAttrs(Context context, @Nullable AttributeSet attrs) {
        if (attrs != null) {
            TypedArray typedArray = context.getTheme().obtainStyledAttributes(
                    attrs,
                    R.styleable.ExploreSetView, 0, 0);
            int textResourceId = typedArray.getResourceId(R.styleable.ExploreSetView_header, 0);
            int backgroundResourceId = typedArray.getResourceId(R.styleable.ExploreSetView_background, 0);
            int headerColorId = typedArray.getResourceId(R.styleable.ExploreSetView_headerColor, 0);
            background.setImageResource(backgroundResourceId);
            header.setText(textResourceId);
            setHeaderColor(headerColorId);
            typedArray.recycle();
        }
    }

    private void initList(ArrayList<Game> games) {
        exploreSetListAdapter = new ExploreSetListAdapter(context, games);
        exploreSetListAdapter.setOnListItemClickListener((pos, v) -> {
            if (onGameClickListener != null)
                onGameClickListener.onClick(exploreSetListAdapter.getGames().get(pos), v);
        });
        list.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        list.setAdapter(exploreSetListAdapter);
    }

    public TextView getHeader() {
        return header;
    }

    public void setHeader(String header) {
        this.header.setText(header);
    }

    public void setHeaderColor(int colorId) {
        header.setTextColor(colorId);
    }

    public void setGamesListContent(ArrayList<Game> games) {
        if (exploreSetListAdapter == null)
            initList(games);
        else {
            exploreSetListAdapter.setGames(games);
            exploreSetListAdapter.notifyDataSetChanged();
        }
    }

    public void addGamesListContent(ArrayList<Game> games) {
        if (exploreSetListAdapter == null) {
            initList((games));
        } else {
            int previousSize = exploreSetListAdapter.getGames().size();
            exploreSetListAdapter.getGames().addAll(games);
            exploreSetListAdapter.notifyItemRangeInserted(previousSize - 1, games.size());
        }
    }

    public void setBackgroundSource(Drawable drawable) {
        background.setImageDrawable(drawable);
    }

    public void setBackgroundSource(Bitmap bitmap) {
        background.setImageBitmap(bitmap);
    }

    public void setBackgroundColor(int color) {
        background.setBackgroundColor(color);
    }

    public void chooseBackgroundFromGamesList() {
        String chosenUrl = null;
        for (Game game : exploreSetListAdapter.getGames()) {
            if (game.getScreenshots() != null && game.getScreenshots().size() > 1) {
                chosenUrl = game.getScreenshots().get(1).getUrl();
                break;
            }
        }
        if (chosenUrl != null) {
            ImageUrlBuilder imageUrlBuilder = new ImageUrlBuilder(chosenUrl);
            Glide.with(context)
                    .load(imageUrlBuilder.setSize(ImageUrlBuilder.ImageSize.screenshot_big).build())
                    .into(background);
        }
    }

    public void setBackgroundTintVisibility(boolean visibility) {
        if (visibility)
            backgroundTint.setVisibility(VISIBLE);
        else {
            backgroundTint.setVisibility(INVISIBLE);
        }
    }

    public void setBackgroundTintAlpha(float alpha) {
        backgroundTint.setAlpha(alpha);
    }

    public void setBackgroundTintColor(int color) {
        backgroundTint.setBackgroundColor(color);
    }
}

File: app/src/main/java/com/soutvoid/gamesproject/ui/screen/main/widgets/exploreset/widget/ExploreSetView.kt
package com.soutvoid.gamesproject.ui.screen.main.widgets.exploreset.widget

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.bumptech.glide.Glide
import com.soutvoid.gamesproject.domain.game.Game
import com.soutvoid.gamesproject.interactor.util.ImageUrlBuilder
import com.soutvoid.gamesproject.ui.screen.main.widgets.exploreset.list.ExploreSetListAdapter
import com.soutvoid.gamesproject.ui.util.OnGameClickListener
import soutvoid.com.gamesproject.R
import java.util.*

class ExploreSetView : FrameLayout {

    private var viewContext: Context? = null

    @BindView(R.id.main_explore_set_container)
    internal var container: ViewGroup? = null
    @BindView(R.id.main_explore_set_background)
    internal var background: ImageView? = null
    @BindView(R.id.main_explore_set_background_tint)
    internal var backgroundTint: View? = null
    @BindView(R.id.main_explore_set_header)
    var header: TextView? = null
        internal set
    @BindView(R.id.main_explore_set_list)
    internal var list: RecyclerView? = null

    private var exploreSetListAdapter: ExploreSetListAdapter? = null

    private var onGameClickListener: OnGameClickListener? = null

    constructor(context: Context) : super(context) {
        init(context, null)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        this.viewContext = context
        View.inflate(context, R.layout.explore_set_view, this)
        ButterKnife.bind(this)
        fillFromAttrs(context, attrs)
    }

    private fun fillFromAttrs(context: Context, attrs: AttributeSet?) {
        if (attrs != null) {
            val typedArray = context.theme.obtainStyledAttributes(
                    attrs,
                    R.styleable.ExploreSetView, 0, 0)
            val textResourceId = typedArray.getResourceId(R.styleable.ExploreSetView_header, 0)
            val backgroundResourceId = typedArray.getResourceId(R.styleable.ExploreSetView_background, 0)
            val headerColorId = typedArray.getResourceId(R.styleable.ExploreSetView_headerColor, 0)
            background!!.setImageResource(backgroundResourceId)
            header!!.setText(textResourceId)
            setHeaderColor(headerColorId)
            typedArray.recycle()
        }
    }

    private fun initList(games: ArrayList<Game>) {
        exploreSetListAdapter = ExploreSetListAdapter(viewContext, games)
        exploreSetListAdapter!!.setOnListItemClickListener { pos, v ->
            if (onGameClickListener != null)
                onGameClickListener!!.onClick(exploreSetListAdapter!!.games[pos], v)
        }
        list!!.layoutManager = LinearLayoutManager(viewContext, LinearLayoutManager.HORIZONTAL, false)
        list!!.adapter = exploreSetListAdapter
    }

    fun setHeader(header: String) {
        this.header!!.text = header
    }

    fun setHeaderColor(colorId: Int) {
        header!!.setTextColor(colorId)
    }

    fun setGamesListContent(games: ArrayList<Game>) {
        if (exploreSetListAdapter == null)
            initList(games)
        else {
            exploreSetListAdapter!!.games = games
            exploreSetListAdapter!!.notifyDataSetChanged()
        }
    }

    fun addGamesListContent(games: ArrayList<Game>) {
        if (exploreSetListAdapter == null) {
            initList(games)
        } else {
            val previousSize = exploreSetListAdapter!!.games.size
            exploreSetListAdapter!!.games.addAll(games)
            exploreSetListAdapter!!.notifyItemRangeInserted(previousSize - 1, games.size)
        }
    }

    fun setBackgroundSource(drawable: Drawable) {
        background!!.setImageDrawable(drawable)
    }

    fun setBackgroundSource(bitmap: Bitmap) {
        background!!.setImageBitmap(bitmap)
    }

    override fun setBackgroundColor(color: Int) {
        background!!.setBackgroundColor(color)
    }

    fun chooseBackgroundFromGamesList() {
        var chosenUrl: String? = null
        for ((_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, screenshots) in exploreSetListAdapter!!.games) {
            if (screenshots != null && screenshots.size > 1) {
                chosenUrl = screenshots[1].getUrl()
                break
            }
        }
        if (chosenUrl != null) {
            val imageUrlBuilder = ImageUrlBuilder(chosenUrl)
            Glide.with(viewContext)
                    .load(imageUrlBuilder.setSize(ImageUrlBuilder.ImageSize.screenshot_big).build())
                    .into(background!!)
        }
    }

    fun setBackgroundTintVisibility(visibility: Boolean) {
        if (visibility)
            backgroundTint!!.visibility = View.VISIBLE
        else {
            backgroundTint!!.visibility = View.INVISIBLE
        }
    }

    fun setBackgroundTintAlpha(alpha: Float) {
        backgroundTint!!.alpha = alpha
    }

    fun setBackgroundTintColor(color: Int) {
        backgroundTint!!.setBackgroundColor(color)
    }

    fun setOnGameClickListener(onGameClickListener: OnGameClickListener) {
        this.onGameClickListener = onGameClickListener
    }
}
